pennylane/qnode.py

```diff
@@ -655,6 +657,34 @@ def construct(self, args, kwargs):
             params = self.qtape.get_parameters(trainable_only=False)
             self.qtape.trainable_params = qml.math.get_trainable_indices(params)
 
+        self.update_device_options()
+
+    def update_device_options(self):
+        """Update the device options once the QNode has been constructed."""
+
+        if self.device.short_name == "default.qubit.torch":
+
+            # Check if we should be using CUDA
+            ops_and_obs = self.qtape.operations + self.qtape.observables
+            any_op_uses_cuda = any(
+                data.is_cuda for op in ops_and_obs for data in op.data if hasattr(data, "is_cuda")
+            )
+
+            if any_op_uses_cuda and self.device._torch_device == "cpu":

```

Comment:
@antalszava I must admit, this feels like a check that shouldn't be in the QNode, but instead be part of the `default.qubit.torch` device.

Instead of passing `torch_device=...` to the Device, could we not move this logic to the device instead? E.g.,

```python
def apply(circuit,...):
    ops_and_obs = circuit.operations + circuit.observables
    any_op_uses_cuda = any(
        data.is_cuda for op in ops_and_obs for data in op.data if hasattr(data, "is_cuda")
    )
    ...
```

This way:

- The QNode avoids having interface and device specific logic
- The Device simply chooses the torch device to use by checking what torch devices the received circuit uses, which feels much cleaner :)

Reply:
Essentially: the `default.qubit.torch` device should check, based on the input circuit, whether it is a GPU circuit. If it _is_, all tensors the device creates should also be on the same device.

Reply:
@josh146 that's a fair idea, frankly, I had this logic in the `execute` method first. The reason why I ended up putting this in the QNode is that the reason for the change came from the QNode itself: at the moment the QNode "silently" creates a `default.qubit.torch` device with the default `torch_device='cpu'` option, while the circuit might be using the GPU. This is only happening because the QNode swaps out `default.qubit` to a `default.qubit.torch` device. When the user explicitly instanties a `default.qubit.torch` device, the responsibility is on them to choose the correct `torch_device` argument matching their circuit (GPU or CPU).

> The QNode avoids having interface and device specific logic

We'll likely have to include some logic sooner or later in the QNode because right now the QNode instantiates all backprop devices with default arguments. It feels like we're offloading the duty of the QNode getting the correct device options onto the devices, just to preserve the workflow of the QNode:

a) creating a device with default options, but then
b) the device itself updates those device options under the hood.

> The Device simply chooses the torch device to use by checking what torch devices the received circuit uses, which feels much cleaner :)

While this sounds good to have, frankly I'd think we're making up for the poor use of the device by the user if we have logic like this in the device.

All in all, I'd argue keeping the logic in the QNode, as that feels like the place where this logic belongs, just because it's a QNode use case. Having said that, there are indeed merits to moving this to the device. :slightly_smiling_face: